Add Newton-Raphson root finder

nilsvu opened the pull request:
## Proposed changes

Adds `RootFinder::newton_raphson` and renames the existing TOMS748 root finder from `find_root_of_function` to `RootFinder::toms748`.

**What breaks:**
- Code that calls `find_root_of_function` must be refactored to `RootFinder::newton_raphson` if derivatives are available, or `RootFinder::toms748` if they are not.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

nilsvu committed:
Add NewtonRaphson root finder

nilsvu commented:
Please add _breaking change_ label.

mar-celine commented on src/NumericalAlgorithms/RootFinding/RootFinder.hpp:
```diff
@@ -2,16 +2,18 @@
 
 /// \file
-/// Declares function find_root_of_function
+/// Declares function for root finding
```
/// \file comment not needed - we'll want a single hpp for each root finder

mar-celine commented on src/NumericalAlgorithms/RootFinding/RootFinder.hpp:
```diff
@@ -47,8 +48,7 @@ double find_root_of_function(const Function& f, const double lower_bound,
   // tolerance
-  // clang-tidy: internal boost warning, can't fix it.
-  auto result = boost::math::tools::toms748_solve(  // NOLINT
+  auto result = boost::math::tools::toms748_solve(
```
Generally do not remove NOLINTs

mar-celine commented on src/NumericalAlgorithms/RootFinding/RootFinder.hpp:
```diff
@@ -106,3 +105,74 @@ DataVector find_root_of_function(const Function& f,
+                                    << std::numeric_limits<double>::digits10
+                                    << " base-10 digits.");
+
+  boost::uintmax_t it = max_iterations;
```
instead of `it`, call this `max_iters`

mar-celine commented on src/NumericalAlgorithms/RootFinding/RootFinder.hpp:
```diff
@@ -106,3 +105,74 @@ DataVector find_root_of_function(const Function& f,
+ *
+ * See the [Boost](http://www.boost.org/) documentation for more details.
+ *
+ * \requires Function f be callable with a `double` and a `size_t`
```
`f` instead of f

mar-celine commented on src/NumericalAlgorithms/RootFinding/RootFinder.hpp:
```diff
@@ -106,3 +105,74 @@ DataVector find_root_of_function(const Function& f,
+ *
+ * See the [Boost](http://www.boost.org/) documentation for more details.
+ *
+ * \requires Function f is invokable with a `double`
```
`f` instead of f (use ticks)

mar-celine commented on src/NumericalAlgorithms/RootFinding/RootFinder.hpp:
```diff
@@ -106,3 +105,74 @@ DataVector find_root_of_function(const Function& f,
+
+  DataVector result_vector{lower_bound.size()};
+  for (size_t i = 0; i < result_vector.size(); ++i) {
+    boost::uintmax_t it = max_iterations;
```
here too

mar-celine commented on src/NumericalAlgorithms/RootFinding/RootFinder.hpp:
```diff
@@ -106,3 +105,74 @@ DataVector find_root_of_function(const Function& f,
+    boost::uintmax_t it = max_iterations;
+    result_vector[i] = boost::math::tools::newton_raphson_iterate(
+        [&f, i](double x) { return f(x, i); }, initial_guess[i], lower_bound[i],
+        upper_bound[i], std::round(std::log2(std::pow(10, digits))), it);
```
put the computation outside of the loop

mar-celine replied:
put the computation on digits outside of the loop

mar-celine replied:
put the computation on digits outside of the loop

mar-celine replied:
put the computation outside of the loop

mar-celine commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_OneDRootFinder.cpp:
```diff
@@ -104,15 +107,15 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.TOMS748RootSolver.DataVector",
-[[noreturn]] SPECTRE_TEST_CASE(
-    "Unit.Numerical.RootFinding.TOMS748RootSolver.RelativeTol.Double",
-    "[NumericalAlgorithms][RootFinding][Unit]") {
+    // [[OutputRegex, The relative tolerance is too small.]]
```
undo this - no spaces here please - also rename the test to secant.
check through the rest of the file as well, there should be no mention of TOMS left after this rename

mar-celine commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_OneDRootFinder.cpp:
```diff
@@ -121,8 +124,102 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.TOMS748RootSolver.DataVector",
+
+// RootFinder::newton_raphson
+
+SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.NewtonRaphson",
```
move test to new file

mar-celine commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_OneDRootFinder.cpp:
```diff
@@ -121,8 +124,102 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.TOMS748RootSolver.DataVector",
+  const double guess = 1.5;
+  const double lower = 1.;
+  const double upper = 2.;
+  const auto f_and_deriv_lambda = [](double x) {
```
noexcept the lambda

mar-celine commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_OneDRootFinder.cpp:
```diff
@@ -121,8 +124,102 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.TOMS748RootSolver.DataVector",
+      RootFinder::newton_raphson(f_and_deriv_free, guess, lower, upper, digits);
+  const auto root_from_functor = RootFinder::newton_raphson(
+      f_and_deriv_functor, guess, lower, upper, digits);
+  CHECK(std::abs(root_from_lambda - correct) < std::pow(10, digits));
```
did you mean 1 over pow(10,digits)?

mar-celine commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_OneDRootFinder.cpp:
```diff
@@ -3,18 +3,21 @@
+std::pair<double, double> f_and_deriv_free(double x) {
+  return std::make_pair(2. - square(x), -2. * x);
+}
+struct FAndDeriv {
```
FuncAndDeriv

mar-celine commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_OneDRootFinder.cpp:
```diff
@@ -121,8 +124,102 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.TOMS748RootSolver.DataVector",
+
+  root = RootFinder::newton_raphson(f_and_deriv_lambda, guess, lower, upper,
+                                    digits);
+  CHECK(std::abs(root - correct) < std::pow(10, digits));
```
check for others of these

mar-celine commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_OneDRootFinder.cpp:
```diff
@@ -121,8 +124,102 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.TOMS748RootSolver.DataVector",
+
+  auto root = RootFinder::newton_raphson(f_and_deriv_lambda, guess, lower,
+                                         upper, digits);
+  /// [double_newton_raphson_root_find]
```
use an example that doesn't find a root at the bounds given

mar-celine commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_OneDRootFinder.cpp:
```diff
@@ -121,8 +124,102 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.TOMS748RootSolver.DataVector",
+
+// [[OutputRegex, The desired accuracy of 100 base-10 digits must be smaller]]
+[[noreturn]] SPECTRE_TEST_CASE(
+    "Unit.Numerical.RootFinding.NewtonRaphson.Digits",
```
also test this for datavector

fmahebert commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_OneDRootFinder.cpp:
```diff
@@ -3,18 +3,21 @@
 
 namespace {
-double f_free(double x) { return 2.0 - x * x; }
+double f_free(double x) { return 2. - square(x); }
```
If you are making this change, maybe make it throughout the file too?

nilsvu commented:
I addressed your comments. Do you actually get a notification about this automatically so that I just push the changes and leave it at that, or should I write a comment like this to make you aware of it?

teukolsky commented:
Renaming TOMS 748 to Secant is a bad idea - it's not a secant method, and is much better than the secant method. The algorithm is actually a high-order inverse interpolation method analogous to Brent's method, although it's claimed to be better than Brent (I haven't verified that myself). I'm happy to leave it's name as TOMS 748. It (or Brent) is the algorithm of choice when derivatives of the function are not easily available.

Sorry I didn't catch this earlier.

kidder commented on src/NumericalAlgorithms/RootFinding/Secant.hpp:
```diff
@@ -97,7 +97,7 @@ DataVector secant(const Function& f, const DataVector& lower_bound,
     // max_iterations after each root find.
     boost::uintmax_t max_iters = max_iterations;
-    auto result = boost::math::tools::toms748_solve(
+    auto result = boost::math::tools::toms748_solve(  // NOLINT
```
add comment  for reason of NOLINT

nilsvu replied:
Actually I don't know the reason why clang-tidy fails for this function, apparently it's an issue somewhere in boost. Indeed also the Travis check for clang-tidy still fails (see build for this PR), despite the `NOLINT`s, could you help me out with that?

The issue appears to be in the `double` version of the `RootFinder::toms748` function. Similar issues in `RootFinder::newton_raphson` I was able to silence with `NOLINT`s, but I can't make it work here for some reason.

nilsdeppe replied:
Clang-tidy fails because of something in Boost. Will looked into this at one point and clang-tidy considers this a "feature" (we don't since we can't fix code inside Boost), so the best we can do is offer an explanation and ignore the build failure. The result of the NOLINT is that rather than pointing into Boost header files it points to our invocation of Boost and so then people can read our comment.

nilsvu replied:
Ok I added comments. Now should I also `NOLINT` all calls to `RootFinder::toms748`?

wthrowe replied:
Just put back the comment you removed earlier in the PR.

nilsvu replied:
Ok done

kidder commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_NewtonRaphson.cpp:
```diff
@@ -0,0 +1,140 @@
+#include "ErrorHandling/Error.hpp"
+#include "NumericalAlgorithms/RootFinding/NewtonRaphson.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "tests/Unit/TestingFramework.hpp"
```
should be first include (i.e. right after the license) (see https://spectre-code.org/writing_unit_tests.html)

nilsdeppe commented on src/NumericalAlgorithms/RootFinding/NewtonRaphson.hpp:
```diff
@@ -0,0 +1,88 @@
+                                       "the machine numeric limit of "
+                                    << std::numeric_limits<double>::digits10
+                                    << " base-10 digits.");
+  auto digits_binary = std::round(std::log2(std::pow(10, digits)));
```
I expect this can be const

nilsdeppe commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_TOMS748.cpp:
```diff
@@ -9,45 +9,46 @@
 
 namespace {
-double f_free(double x) { return 2.0 - x * x; }
+double f_free(double x) { return 2. - square(x); }
```
It looks like the majority of this file uses `2.0`, could you make the whole file consistent? :)

nilsdeppe reviewed (dismissed):
I'm happy for you to rebase and squash

nilsdeppe reviewed (changes requested):
You have an issue with your doxygen documentation

nilsvu commented:
Should be fixed, let's wait for Travis to catch up :)

**Edit:** Only boost-internal `clang-tidy` warnings left.
